modules/services/fnott.nix

```diff
@@ -0,0 +1,180 @@
+
+  attrPaths' = attrs:
+    collect isList (mapAttrsRecursive (path: value: path) attrs);
+  optionMapping = {
+    output = [ "main" "output" ];
+    anchor = [ "main" "anchor" ];
+    playSound = [ "main" "play-sound" ];
+    maxIconSize = [ "main" "max-icon-size" ];
+    iconTheme = [ "main" "icon-theme" ];
+    maxHeight = [ "main" "max-height" ];
+    minWidth = [ "main" "min-width" ];
+    maxWidth = [ "main" "max-width" ];
+  };
+
+in {
+  options = {
+    services.fnott = {
+      enable = mkEnableOption ''
+        fnott, a lightweight Wayland notification daemon for wlroots-based compositors.
+      '';
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.fnott;
+        defaultText = literalExample "pkgs.fnott";
+        description = "Package providing <command>fnott</command>.";
+      };
+
+      output = mkOption {
+        type = types.str;
+        description = ''
+          Show notifications on the specified output. If empty, notifications will
+          appear on the focused output. Requires the compositor to support the Wayland
+          protocol xdg-output-unstable-v1 version 2.
+
+          On
+          <citerefentry>
+            <refentrytitle>sway</refentrytitle>
+            <manvolnum>1</manvolnum>
+          </citerefentry>,
+          you can use <literal>swaymsg -t get_outputs</literal> to get a list of
+          avalaible outputs.
+        '';
+      };
+
+      anchor = mkOption {
+        type =
+          types.enum [ "top-right" "top-left" "bottom-right" "bottom-left" ];
+        description = ''
+          Show notifications at the specified position on the output.
+          Supported values are top-right, top-left, bottom-right,
+          and bottom-left.
+        '';
+      };
+
+      playSound = mkOption {
+        type = types.str;
+        defaultText =
+          literalExample "\${pkgs.alsa-utils}/bin/aplay \${filename}";
+        description = ''
+          Command to execute to play notification sounds.
+          ''${filename} will be expanded to the path
+          to the audio file to play.
+        '';
+      };
+
+      maxIconSize = mkOption {
+        type = types.int;
+        description = ''
+          Maximum icon size, in pixels. Icons larger than this will be scaled down.
+        '';
+      };
+
+      iconTheme = mkOption {
+        type = types.str;
+        description = ''
+          Icon theme to use when a notification has requested an non-embedded icon.
+        '';
+      };
+
+      maxHeight = mkOption {
+        type = types.int;
+        description = "Maximum notification height, in pixels.";
+      };
+
+      minWidth = mkOption {
+        type = types.int;
+        description = "Minimum notification width, in pixels.";
+      };
+
+      maxWidth = mkOption {
+        type = types.int;
+        description = "Maximum notification width, in pixels.";
+      };
+
+      settings = mkOption {
```

Comment:
I think you'll want to make this a [freeform module](https://nixos.org/manual/nixos/stable/index.html#sec-freeform-modules). That should allow you to remove the special handling with `optionMapping`.

Reply:
initially I was going to make it a freeform module, but I was planning to submit a patch upstream to allow `[main]` as also the default section like in foot, thus the entire configuration can then be a structural `settings`. I think I was aiming for either all or nothing at the time, I agree this is better way to go about it.

Reply:
when implementing the freeform module, I am not sure how to define the module such that its submodule options do not need to be defined, i.e isDefined == false, while still allowing the upper level module to evaluate as mergeDefinitions will throw an error. 

If I add mkIf guards to individual submodule options, I end up in the same situation as it is currently, albeit a bit cleaner.

sorry for the long messages, I hope you can understand what I am trying to convey?